experimental/cuda2/cuda_driver.c

```diff
@@ -0,0 +1,448 @@
+           "%02x%02x%02x%02x-"
+           "%02x%02x-"
+           "%02x%02x-"
+           "%02x%02x-"
+           "%02x%02x%02x%02x%02x%02x",
+           (uint8_t)device_uuid.bytes[0], (uint8_t)device_uuid.bytes[1],
+           (uint8_t)device_uuid.bytes[2], (uint8_t)device_uuid.bytes[3],
+           (uint8_t)device_uuid.bytes[4], (uint8_t)device_uuid.bytes[5],
+           (uint8_t)device_uuid.bytes[6], (uint8_t)device_uuid.bytes[7],
+           (uint8_t)device_uuid.bytes[8], (uint8_t)device_uuid.bytes[9],
+           (uint8_t)device_uuid.bytes[10], (uint8_t)device_uuid.bytes[11],
+           (uint8_t)device_uuid.bytes[12], (uint8_t)device_uuid.bytes[13],
+           (uint8_t)device_uuid.bytes[14], (uint8_t)device_uuid.bytes[15]);
+  buffer_ptr += iree_string_view_append_to_buffer(
+      iree_make_string_view(device_path_str,
+                            IREE_ARRAYSIZE(device_path_str) - 1),
+      &out_device_info->path, (char*)buffer_ptr);
+
+  iree_string_view_t device_name_str =
+      iree_make_string_view(device_name, strlen(device_name));
+  buffer_ptr += iree_string_view_append_to_buffer(
+      device_name_str, &out_device_info->name, (char*)buffer_ptr);
+
+  *out_buffer_ptr = buffer_ptr;
+  return iree_ok_status();
+}
+
+// Returns true if the device meets all the required capabilities.
+static bool iree_hal_cuda2_is_valid_device(iree_hal_cuda2_driver_t* driver,
+                                           CUdevice device) {
+  return true;
+}
+
+static iree_status_t iree_hal_cuda2_driver_query_available_devices(
+    iree_hal_driver_t* base_driver, iree_allocator_t host_allocator,
+    iree_host_size_t* out_device_info_count,
+    iree_hal_device_info_t** out_device_infos) {
+  IREE_ASSERT_ARGUMENT(base_driver);
+  IREE_ASSERT_ARGUMENT(out_device_info_count);
+  IREE_ASSERT_ARGUMENT(out_device_infos);
+  iree_hal_cuda2_driver_t* driver = iree_hal_cuda2_driver_cast(base_driver);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // Ensure CUDA is initialized before querying it.
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(z0, iree_hal_cuda2_init(driver));
+
+  // Query the number of available CUDA devices.
+  int device_count = 0;
+  CUDA_RETURN_AND_END_ZONE_IF_ERROR(z0, &driver->cuda_symbols,
+                                    cuDeviceGetCount(&device_count),
+                                    "cuDeviceGetCount");
+
+  // Allocate the return infos and populate with the devices.
+  iree_hal_device_info_t* device_infos = NULL;
+  iree_host_size_t total_size =
+      device_count * (sizeof(iree_hal_device_info_t) +
+                      IREE_HAL_CUDA_MAX_DEVICE_NAME_LENGTH * sizeof(char));
+  iree_status_t status =
+      iree_allocator_malloc(host_allocator, total_size, (void**)&device_infos);
+
+  int valid_device_count = 0;
+  if (iree_status_is_ok(status)) {
+    uint8_t* buffer_ptr =
+        (uint8_t*)device_infos + device_count * sizeof(iree_hal_device_info_t);
+    for (iree_host_size_t i = 0; i < device_count; ++i) {
+      CUdevice device;
```

Comment:
always default initialize values in C (= 0)
```suggestion
      CUdevice device = 0;
```

even if not "required" when debugging broken error handling or stepping through this makes things much easier to track - if stepping across the cuDeviceGet call a dev will see a predictable 0, make the call, and see the result (hopefully non-zero) - if it's uninitialized that makes things trickier.

Reply:
Good points!